### PR TITLE
Fixes perplexity computation (issue https://github.com/NVIDIA/NeMo/issues/1289)

### PR DESCRIPTION
**Pull request by PeganovAnton**
* Add perplexity draft

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

* Add perplexity metric

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

* Add docstrings and copyright, fix comments

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

* Fix code style

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

* Fix code style

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

* Improve docstring

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

* Fix docstring hrefs

Signed-off-by: PeganovAnton <peganoff2@mail.ru>

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/bert_lm_model.py [hunk 1]
```diff
@@ -153,19 +155,21 @@ def validation_step(self, batch, batch_idx):
...
             loss = mlm_loss
         else:
-            nsp_loss = self.nsp_loss(logits=logits[1], labels=labels)
+            nsp_loss = self.nsp_loss(logits=nsp_logits, labels=labels)
 
             loss = self.agg_loss(loss_1=mlm_loss, loss_2=nsp_loss)
-        perplexity = self.perplexity_metric(mlm_loss)
-        tensorboard_logs = {"val_loss": loss, "perplexity": perplexity}
-        return {"val_loss": loss, "log": tensorboard_logs}
+        self.validation_perplexity(logits=mlm_log_probs)
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/language_modeling/bert_lm_model.py:170]: so it looks like here we are passing log_probs instead of logits. Is this a bug?

**Review by okuchaiev — CHANGES_REQUESTED** [review]

File: nemo/collections/nlp/metrics/perplexity.py [hunk 2]
```diff
@@ -1,38 +0,0 @@
...
-    """
-    This metric computes the perplexity given the language model loss.
-    """
-
-    def __init__(self, dist_sync_on_step=False):
-        super().__init__(dist_sync_on_step=dist_sync_on_step)
-        self.add_state('perplexity', default=torch.tensor(0), dist_reduce_fx='mean', persistent=False)
-
-    def update(self, loss: torch.Tensor):
-        self.perplexity = torch.exp(loss)
```
Comment (okuchaiev) [inline nemo/collections/nlp/metrics/perplexity.py:35]: for certain losses, such as: -p*log(p) this looks correct, no?

File: nemo/collections/nlp/models/language_modeling/bert_lm_model.py (hunk 1 above)
Comment (PeganovAnton) [reply nemo/collections/nlp/models/language_modeling/bert_lm_model.py:170]: In multilabel case log probs are logits normalized via `log_probs = log(logits /  exp(logits).sum())`, so log probs are logits but logits are usually not log probs. If you apply softmax to log probs, you will have probs.

File: nemo/collections/nlp/metrics/perplexity.py (hunk 2 above)
Comment (PeganovAnton) [reply nemo/collections/nlp/metrics/perplexity.py:35]: Cross-entropy loss is calculated using predicted probabilities and target probabilities.
`cross_entropy_loss = -(target_probs * log(predicted_probs)).sum()`
Perplexity exponentiates not cross-entropy but entropy. Entropy is not a loss because it does not compare predictions to targets. Entropy and perplexity just show how confident is a model about its predictions.

**Comment by titu1994** [comment]
@PeganovAnton Could you add the Perplexity metric to the Transformer based language model class in NeMo ?

**Review by titu1994 — COMMENTED** [review]
Could you add the Perplexity metric to the Transformer based language model class in NeMo ?

**Review by okuchaiev — APPROVED** [review]
